### ISSUE TITLE
Integration Events

### ISSUE THREAD

**Issue by Dismissile** [dotnet/docs#12469]
"As described earlier, when you use event-based communication, a microservice publishes an event when something notable happens, such as when it updates a business entity. Other microservices subscribe to those events."

"The integration events can be defined at the application level of each microservice, so they are decoupled from other microservices, in a way comparable to how ViewModels are defined in the server and client. What is not recommended is sharing a common integration events library across multiple microservices; doing that would be coupling those microservices with a single event definition data library. You do not want to do that for the same reasons that you do not want to share a common domain model across multiple microservices: microservices must be completely autonomous."

These two statements almost seem at odds with one another. If the events should be defined within the microservice itself, how can another microservice subscribe to those events without duplicating the event structure entirely?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b037f198-3ee2-28f4-4d93-0999e5d3d3ef
* Version Independent ID: 9bd96308-c726-fcee-850e-a38fd22ff555
* Content: [Implementing event-based communication between microservices (integration events)](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/multi-container-microservice-net-applications/integration-event-based-microservice-communications#feedback)
* Content Source: [docs/architecture/microservices/multi-container-microservice-net-applications/integration-event-based-microservice-communications.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/multi-container-microservice-net-applications/integration-event-based-microservice-communications.md)
* Product: **dotnet**
* Technology: **dotnet-ebooks**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

**Comment by jwisener** [comment]
Agreed!

**Comment by gladium-software** [comment]
Hi Dismissile and Jwisener,

It might seem at odds, and if you’re taking DRY to the fullest extent you’ll probably not like what I tend to implement. 

As stated in the article, to keep the services fully decoupled you don’t want to share a Contract library between the services - doing so intrinsically ties the services together - got the T-Shirt on that one. 

What you’re best doing is thinking as though you’re not in control of the development of both services and that you’re simply receiving a webhook event or making a rest call to a third party. In this case you would probably define a set of classes that represent the calls on both sides. Yes, this does mean that you’ll have what looks like repeated code, eg a class representing an event in the publishing service and a class representing the event in the receiving class but that’s ok, both services are within their own boundaries.

If you don’t do this then when it comes to releasing a new/updated contract/event you have to look at releasing the associated services - therefore losing the benefit of the decoupling. 

My 2c

**Comment by jwisener** [comment]
My apologies it’s been close to 4 months since i posted that comment and completely forgot about it. I am not for copying and pasting that duplicate class definitions across our various projects. For me, Integration means exactly what it’s name implies, an agreement between two or more systems, in order for that to happen they have to agree upon a contract, I just prefer to not copy and paste it everywhere. I tend to favour the concept of fail as quickly as possible.  If a contract is changed, then I like to see the compile time errors and fail as quickly as possible, vs deploying and then realising ‘oh we forget to change subsystem x, y and z’. Even if I had two services subscribing to - 3rd party endpoint or library I must adhere to that contract. Don’t get me wrong, not everything is always ‘black and white’. But for our integration events we choose to define them in this manner (a integration core library). If something needs to be extended we then look at the various ways we need to accommodate the addition/change, doing it in this way helps analyse the impact better.

**Comment by gladium-software** [comment]
Hey - and this is why I love software development, if you find something that works and works well, stick with it!

I should have chosen my words a little more carefully as I didn’t want to suggest another way is wrong - purely giving an insight on how I've implemented this before. 

I completely agree on the contract side of things i.e. the two services need a contract. When defining the contract I tend to think of them more as a specification that needs to be coded to. So the publisher provides a specification of what’s going to be published and the consumer codes to that specification.

I work on platforms that are polyglot, although mostly .net there’s a good mixture of C++, Java, Go etc, so I take the approach that a service could be replaced with another language at some point - even if the services are initially .net. By not relying on a contract library I’m not bound by the technology. 

When it comes to failing fast, again, completely agree. What I tend to do here is take a two phased approach; the first being that the changes to the specification are coded to and then unit tested within the bounds of the service, and the second is a reliance on integration testing.

**Comment by mvelosop** [comment]
Hi @Dismissile, @jwisener, @gladium-software, thanks for your feedback.

This is always an interesting, and recurring, discussion.

I used to have the same opinion as @Dismissile and @jwisener, but then, just recently, I started working in a large microservices app with about 12 development teams and close to 70 developers and my point of view has changed completely.

In that context, things are just completely different and now I understand @gladium-software opinion.

In the end, as @gladium-software stated, if it works for you, that is, your app, your team(s), your context, keep using it, but just as important, keep in mind that scale can change the context dramatically. 

So, if you face something like this is affecting your development team, be ready to pivot and try something else.

There's a discussion with @CESARDELATORRE about this topic in [this issue on the eShopOnContainers repo](https://github.com/dotnet-architecture/eShopOnContainers/issues/724) that you might find interesting.

Just my two cents.

I'm closing this issue now, but feel free to comment, will reopen if needed.
